packages/mattermost-redux/src/utils/apps.ts

```diff
@@ -1,116 +1,189 @@
-// - Inherit app calls
-// - Inherit app ids
-// - Trim invalid bindings (do not have an app call or app id at the leaf)
-export function fillAndTrimBindingsInformation(binding?: AppBinding) {
+import {AppBinding, AppField, AppForm} from '../types/apps';
+import {AppBindingLocations, AppFieldTypes} from '../constants/apps';
+
+export function cleanBinding(binding: AppBinding, topLocation: string, depth: number) {
     if (!binding) {
         return;
     }
 
-    binding.bindings?.forEach((b) => {
-        // Propagate id down if not defined
+    const toRemove: number[] = [];
+    const usedLabels: {[label: string]: boolean} = {};
+    binding.bindings?.forEach((b, i) => {
+        // Inheritance and defaults
+        if (!b.call) {
+            b.call = binding.call;
+        }
+
+        if (b.form) {
+            cleanForm(b.form);
+        } else {
+            b.form = binding.form;
+        }
+
         if (!b.app_id) {
             b.app_id = binding.app_id;
         }
 
-        // Compose location
+        if (!b.label) {
+            b.label = b.location || '';
+        }
+
         b.location = binding.location + '/' + b.location;
 
-        // Propagate call down if not defined
-        if (!b.call) {
-            b.call = binding.call;
+        // Validation
+        if (!b.label) {
+            toRemove.unshift(i);
+            return;
         }
 
-        fillAndTrimBindingsInformation(b);
-    });
-
-    // Trim branches without app_id
-    if (!binding.app_id) {
-        binding.bindings = binding.bindings?.filter((v) => v.app_id);
-    }
+        switch (topLocation) {
+        case AppBindingLocations.COMMAND: {
+            if (b.label.match(/ |\t/)) {
+                toRemove.unshift(i);
+                return;
+            }
+
+            if (usedLabels[b.label]) {
+                toRemove.unshift(i);
+                return;
+            }
+            break;
+        }
+        case AppBindingLocations.CHANNEL_HEADER_ICON: {
+            // First level of channel header icons must have an icon to show as the icon
+            if (!b.icon && depth === 0) {
+                toRemove.unshift(i);
+                return;
+            }
+            break;
+        }
+        }
 
-    // Trim branches without calls
-    if (!binding.call) {
-        binding.bindings = binding.bindings?.filter((v) => v.call);
-    }
+        if (b.bindings?.length) {
+            cleanBinding(b, topLocation, depth + 1);
+
+            // Remove invalid branches
+            if (!b.bindings?.length) {
+                toRemove.unshift(i);
+                return;
+            }
+        } else {
+            // Remove leaves without a call
+            if (!b.call) {
+                toRemove.unshift(i);
+                return;
+            }
+
+            // Remove leaves without app id
+            if (!b.app_id) {
+                toRemove.unshift(i);
+                return;
+            }
+        }
```

Comment:
```suggestion
        } else {
            // Remove leaves without a call or app id
            if (!b.call || !b.app_id) {
                toRemove.unshift(i);
                return;
            }
        }
```
0/5 You could combine the two

Reply:
there are quite a few cases where the `if` statements could be combined to share the common `toRemove.unshift(i)` calls. This is non-blocking and just 0/5 because it's a personal choice :)

Reply:
I preferred being a bit more verbose here, since we may add or remove more rules here eventually.